src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -196,13 +249,26 @@ private async Task ProcessQueueAsync()
                         // will be sent back to the client but they can also be captured via HandleNonMutatingRequestError,
                         // though these errors don't put us into a bad state as far as the rest of the queue goes.
                         // Furthermore we use Task.Run here to protect ourselves against synchronous execution of work
-                        // blocking the request queue for longer periods of time (it enforces parallelizabilty).
-                        _ = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+                        // blocking the request queue for longer periods of time (it enforces parallelizability).
+                        var pendingTask = WrapStartRequestTaskAsync(Task.Run(() => work.StartRequestAsync(context, cancellationToken), cancellationToken), rethrowExceptions: false);
+
+                        if (CancelInProgressWorkUponMutatingRequest)
+                        {
+                            pendingTasks.TryAdd(pendingTask, cancellationTokenSource);
+
+                            _ = pendingTask.ContinueWith(t =>
+                            {
+                                if (pendingTasks.TryRemove(t, out var pendingCancellationTokenSource))
+                                {
+                                    pendingCancellationTokenSource?.Dispose();
+                                }
+                            }, TaskScheduler.Default);
+                        }
                     }
                 }
                 catch (OperationCanceledException ex) when (ex.CancellationToken == queueItem.cancellationToken)
```

Comment:
i don't trust this catch anymore.  we may be passing in a different CT when we execute the task.  so this may no longer apply.

Reply:
The ThrowIfCancellationRequested code hits this (I've seen it hit when I was debugging). The reason that statement was added was becuase the CreateLinkedTokenSource call will throw an ObjectDisposedException if either of the CancellationTokens it is passed are already cancelled.

It's a bit confusing with all the cancellation token sources being thrown around in this code, can you elaborate a bit on your concern and how you think we should address it?

Reply:
i'd just remove the `when` clause altogether.  basically.  let any cancellation get here.  eitehr we continue the loop, or the loop cancels if its actuallly shutdown.